File: examples/php/TragWeb3.php
<?php

require_once 'vendor/autoload.php';

use Web3\Web3;
use Web3\Contract;
use Web3\Utils;
use Web3\Providers\HttpProvider;
use Web3\RequestManagers\HttpRequestManager;

/**
 * TRAG Token Web3 PHP Client
 *
 * This class provides methods to interact with the TRAG token contract
 * on Binance Smart Chain using Web3 PHP library.
 */
class TragToken
{
    private $web3;
    private $contract;
    private $contractAddress;
    private $contractAbi;
    private $privateKey;
    private $account;

    // TRAG Token Configuration
    const CONTRACT_ADDRESS = '0x7Cc723dE7fBDb6B06d6628E259e6B8c62673BF1C';
    const BSC_RPC_URL = 'https://bsc-dataseed1.binance.org/';
    const TOKEN_DECIMALS = 6;

    // Simplified Contract ABI
    const CONTRACT_ABI = '[
        {
            "inputs": [],
            "name": "name",
            "outputs": [{"internalType": "string", "name": "", "type": "string"}],
            "stateMutability": "view",
            "type": "function"
        },
        {
            "inputs": [],
            "name": "symbol",
            "outputs": [{"internalType": "string", "name": "", "type": "string"}],
            "stateMutability": "view",
            "type": "function"
        },
        {
            "inputs": [],
            "name": "decimals",
            "outputs": [{"internalType": "uint8", "name": "", "type": "uint8"}],
            "stateMutability": "view",
            "type": "function"
        },
        {
            "inputs": [],
            "name": "totalSupply",
            "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
            "stateMutability": "view",
            "type": "function"
        },
        {
            "inputs": [{"internalType": "address", "name": "account", "type": "address"}],
            "name": "balanceOf",
            "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
            "stateMutability": "view",
            "type": "function"
        },
        {
            "inputs": [
                {"internalType": "address", "name": "recipient", "type": "address"},
                {"internalType": "uint256", "name": "amount", "type": "uint256"}
            ],
            "name": "transfer",
            "outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "inputs": [
                {"internalType": "address", "name": "owner", "type": "address"},
                {"internalType": "address", "name": "spender", "type": "address"}
            ],
            "name": "allowance",
            "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
            "stateMutability": "view",
            "type": "function"
        },
        {
            "inputs": [
                {"internalType": "address", "name": "spender", "type": "address"},
                {"internalType": "uint256", "name": "amount", "type": "uint256"}
            ],
            "name": "approve",
            "outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
            "stateMutability": "nonpayable",
            "type": "function"
        }
    ]';

    /**
     * Constructor
     *
     * @param string|null $privateKey Private key for transactions (optional)
     */
    public function __construct($privateKey = null)
    {
        // Initialize Web3 with BSC provider
        $this->web3 = new Web3(new HttpProvider(new HttpRequestManager(self::BSC_RPC_URL)));

        $this->contractAddress = self::CONTRACT_ADDRESS;
        $this->contractAbi = json_decode(self::CONTRACT_ABI, true);

        // Initialize contract
        $this->contract = new Contract($this->web3->provider, $this->contractAbi);
        $this->contract->at($this->contractAddress);

        // Setup account if private key provided
        if ($privateKey) {
            $this->privateKey = $privateKey;
            // Account setup would require additional Web3 account handling
        }
    }

    /**
     * Get token information
     *
     * @return array Token information
     * @throws Exception
     */
    public function getTokenInfo()
    {
        $result = [];

        // Get name
        $this->contract->call('name', [], function ($err, $res) use (&$result) {
            if ($err !== null) {
                throw new Exception('Failed to get token name: ' . $err->getMessage());
            }
            $result['name'] = $res[0];
        });

        // Get symbol
        $this->contract->call('symbol', [], function ($err, $res) use (&$result) {
            if ($err !== null) {
                throw new Exception('Failed to get token symbol: ' . $err->getMessage());
            }
            $result['symbol'] = $res[0];
        });

        // Get decimals
        $this->contract->call('decimals', [], function ($err, $res) use (&$result) {
            if ($err !== null) {
                throw new Exception('Failed to get token decimals: ' . $err->getMessage());
            }
            $result['decimals'] = intval($res[0]->toString());
        });

        // Get total supply
        $this->contract->call('totalSupply', [], function ($err, $res) use (&$result) {
            if ($err !== null) {
                throw new Exception('Failed to get total supply: ' . $err->getMessage());
            }
            $totalSupply = $res[0];
            $result['totalSupply'] = $this->fromWei($totalSupply, self::TOKEN_DECIMALS);
            $result['totalSupplyRaw'] = $totalSupply->toString();
        });

        $result['contractAddress'] = $this->contractAddress;

        return $result;
    }

    /**
     * Get token balance for an address
     *
     * @param string $address The address to check
     * @return array Balance information
     * @throws Exception
     */
    public function getBalance($address)
    {
        if (!Utils::isAddress($address)) {
            throw new Exception('Invalid address format');
        }

        $result = [];

        $this->contract->call('balanceOf', [$address], function ($err, $res) use (&$result) {
            if ($err !== null) {
                throw new Exception('Failed to get balance: ' . $err->getMessage());
            }

            $balance = $res[0];
            $result['raw'] = $balance->toString();
            $result['formatted'] = $this->fromWei($balance, self::TOKEN_DECIMALS);
        });

        return $result;
    }

    /**
     * Get allowance for spender from owner
     *
     * @param string $owner Owner address
     * @param string $spender Spender address
     * @return array Allowance information
     * @throws Exception
     */
    public function getAllowance($owner, $spender)
    {
        if (!Utils::isAddress($owner) || !Utils::isAddress($spender)) {
            throw new Exception('Invalid address format');
        }

        $result = [];

        $this->contract->call('allowance', [$owner, $spender], function ($err, $res) use (&$result) {
            if ($err !== null) {
                throw new Exception('Failed to get allowance: ' . $err->getMessage());
            }

            $allowance = $res[0];
            $result['raw'] = $allowance->toString();
            $result['formatted'] = $this->fromWei($allowance, self::TOKEN_DECIMALS);
        });

        return $result;
    }

    /**
     * Transfer tokens (requires private key)
     *
     * @param string $to Recipient address
     * @param string $amount Amount to transfer
     * @return array Transaction result
     * @throws Exception
     */
    public function transfer($to, $amount)
    {
        if (!$this->privateKey) {
            throw new Exception('Private key required for transfers');
        }

        if (!Utils::isAddress($to)) {
            throw new Exception('Invalid recipient address');
        }

        $amountWei = $this->toWei($amount, self::TOKEN_DECIMALS);

        // This would require additional implementation for transaction signing
        // and sending with the private key
        throw new Exception('Transfer function requires additional Web3 account implementation');
    }

    /**
     * Approve tokens for spending (requires private key)
     *
     * @param string $spender Spender address
     * @param string $amount Amount to approve
     * @return array Transaction result
     * @throws Exception
     */
    public function approve($spender, $amount)
    {
        if (!$this->privateKey) {
            throw new Exception('Private key required for approvals');
        }

        if (!Utils::isAddress($spender)) {
            throw new Exception('Invalid spender address');
        }

        $amountWei = $this->toWei($amount, self::TOKEN_DECIMALS);

        // This would require additional implementation for transaction signing
        // and sending with the private key
        throw new Exception('Approve function requires additional Web3 account implementation');
    }

    /**
     * Convert from wei to human readable format
     *
     * @param mixed $value Wei value
     * @param int $decimals Token decimals
     * @return string Formatted value
     */
    private function fromWei($value, $decimals)
    {
        $divisor = bcpow('10', $decimals);
        return bcdiv($value->toString(), $divisor, $decimals);
    }

    /**
     * Convert from human readable format to wei
     *
     * @param string $value Human readable value
     * @param int $decimals Token decimals
     * @return string Wei value
     */
    private function toWei($value, $decimals)
    {
        $multiplier = bcpow('10', $decimals);
        return bcmul($value, $multiplier);
    }
}

// Usage examples
function main()
{
    try {
        // Initialize without private key for read-only operations
        $trag = new TragToken();

        // Get token information
        echo "Token Info:\n";
        $tokenInfo = $trag->getTokenInfo();
        foreach ($tokenInfo as $key => $value) {
            echo "  $key: $value\n";
        }

        // Check balance
        echo "\nBalance Check:\n";
        $balance = $trag->getBalance('0x742d35cc6564c0532d7e2d9db08d8d37b4b4e6c8');
        echo "  Balance: " . $balance['formatted'] . " TRAG\n";

        // For transactions, initialize with private key
        // $tragWithKey = new TragToken('your_private_key_here');
        // $transferResult = $tragWithKey->transfer('recipient_address', '100');

    } catch (Exception $e) {
        echo "Error: " . $e->getMessage() . "\n";
    }
}

// Run examples if this file is executed directly
if (basename(__FILE__) == basename($_SERVER['PHP_SELF'])) {
    main();
}

?>

File: examples/go/trag_web3.go
package main

import (
	"context"
	"crypto/ecdsa"
	"fmt"
	"log"
	"math/big"
	"strings"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/accounts/abi/bind"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/ethclient"
)

// TRAG Token Configuration
const (
	ContractAddress = "0x7Cc723dE7fBDb6B06d6628E259e6B8c62673BF1C"
	BSCRPCUrl       = "https://bsc-dataseed1.binance.org/"
	TokenDecimals   = 6
)

// Contract ABI (simplified for examples)
const ContractABI = `[
	{
		"inputs": [],
		"name": "name",
		"outputs": [{"internalType": "string", "name": "", "type": "string"}],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [],
		"name": "symbol",
		"outputs": [{"internalType": "string", "name": "", "type": "string"}],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [],
		"name": "decimals",
		"outputs": [{"internalType": "uint8", "name": "", "type": "uint8"}],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [],
		"name": "totalSupply",
		"outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [{"internalType": "address", "name": "account", "type": "address"}],
		"name": "balanceOf",
		"outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [
			{"internalType": "address", "name": "recipient", "type": "address"},
			{"internalType": "uint256", "name": "amount", "type": "uint256"}
		],
		"name": "transfer",
		"outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"inputs": [
			{"internalType": "address", "name": "owner", "type": "address"},
			{"internalType": "address", "name": "spender", "type": "address"}
		],
		"name": "allowance",
		"outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [
			{"internalType": "address", "name": "spender", "type": "address"},
			{"internalType": "uint256", "name": "amount", "type": "uint256"}
		],
		"name": "approve",
		"outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
		"stateMutability": "nonpayable",
		"type": "function"
	}
]`

// TragToken represents the TRAG token client
type TragToken struct {
	client         *ethclient.Client
	contractAddr   common.Address
	contractABI    abi.ABI
	privateKey     *ecdsa.PrivateKey
	publicAddress  common.Address
}

// TokenInfo holds basic token information
type TokenInfo struct {
	Name            string
	Symbol          string
	Decimals        uint8
	TotalSupply     *big.Int
	ContractAddress string
}

// Balance represents token balance
type Balance struct {
	Raw       *big.Int
	Formatted *big.Float
}

// NewTragToken creates a new TRAG token client
func NewTragToken(privateKeyHex string) (*TragToken, error) {
	// Connect to BSC network
	client, err := ethclient.Dial(BSCRPCUrl)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to BSC network: %v", err)
	}

	// Parse contract address
	contractAddr := common.HexToAddress(ContractAddress)

	// Parse ABI
	contractABI, err := abi.JSON(strings.NewReader(ContractABI))
	if err != nil {
		return nil, fmt.Errorf("failed to parse contract ABI: %v", err)
	}

	t := &TragToken{
		client:       client,
		contractAddr: contractAddr,
		contractABI:  contractABI,
	}

	// Setup private key if provided
	if privateKeyHex != "" {
		privateKey, err := crypto.HexToECDSA(privateKeyHex)
		if err != nil {
			return nil, fmt.Errorf("invalid private key: %v", err)
		}

		publicKey := privateKey.Public()
		publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
		if !ok {
			return nil, fmt.Errorf("failed to get public key")
		}

		t.privateKey = privateKey
		t.publicAddress = crypto.PubkeyToAddress(*publicKeyECDSA)
	}

	return t, nil
}

// GetTokenInfo retrieves basic token information
func (t *TragToken) GetTokenInfo() (*TokenInfo, error) {
	var (
		name        string
		symbol      string
		decimals    uint8
		totalSupply *big.Int
	)

	// Call contract methods
	callOpts := &bind.CallOpts{Context: context.Background()}

	// Get name
	result := []interface{}{&name}
	err := t.client.CallContract(context.Background(), t.buildCallMsg("name"), nil)
	if err == nil {
		err = t.contractABI.UnpackIntoInterface(&result, "name", result)
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get token name: %v", err)
	}

	// Get symbol
	result = []interface{}{&symbol}
	err = t.client.CallContract(context.Background(), t.buildCallMsg("symbol"), nil)
	if err == nil {
		err = t.contractABI.UnpackIntoInterface(&result, "symbol", result)
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get token symbol: %v", err)
	}

	// Get decimals
	result = []interface{}{&decimals}
	err = t.client.CallContract(context.Background(), t.buildCallMsg("decimals"), nil)
	if err == nil {
		err = t.contractABI.UnpackIntoInterface(&result, "decimals", result)
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get token decimals: %v", err)
	}

	// Get total supply
	result = []interface{}{&totalSupply}
	err = t.client.CallContract(context.Background(), t.buildCallMsg("totalSupply"), nil)
	if err == nil {
		err = t.contractABI.UnpackIntoInterface(&result, "totalSupply", result)
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get total supply: %v", err)
	}

	return &TokenInfo{
		Name:            name,
		Symbol:          symbol,
		Decimals:        decimals,
		TotalSupply:     totalSupply,
		ContractAddress: ContractAddress,
	}, nil
}

// GetBalance retrieves token balance for an address
func (t *TragToken) GetBalance(address string) (*Balance, error) {
	addr := common.HexToAddress(address)

	// Pack method call
	data, err := t.contractABI.Pack("balanceOf", addr)
	if err != nil {
		return nil, fmt.Errorf("failed to pack balanceOf call: %v", err)
	}

	// Make call
	result, err := t.client.CallContract(context.Background(), ethereum.CallMsg{
		To:   &t.contractAddr,
		Data: data,
	}, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to call balanceOf: %v", err)
	}

	// Unpack result
	balance := new(big.Int)
	err = t.contractABI.UnpackIntoInterface(&[]interface{}{balance}, "balanceOf", result)
	if err != nil {
		return nil, fmt.Errorf("failed to unpack balance: %v", err)
	}

	// Convert to decimal
	decimals := big.NewInt(int64(TokenDecimals))
	divisor := new(big.Int).Exp(big.NewInt(10), decimals, nil)
	formatted := new(big.Float).Quo(new(big.Float).SetInt(balance), new(big.Float).SetInt(divisor))

	return &Balance{
		Raw:       balance,
		Formatted: formatted,
	}, nil
}

// GetAllowance retrieves allowance for spender from owner
func (t *TragToken) GetAllowance(owner, spender string) (*Balance, error) {
	ownerAddr := common.HexToAddress(owner)
	spenderAddr := common.HexToAddress(spender)

	// Pack method call
	data, err := t.contractABI.Pack("allowance", ownerAddr, spenderAddr)
	if err != nil {
		return nil, fmt.Errorf("failed to pack allowance call: %v", err)
	}

	// Make call
	result, err := t.client.CallContract(context.Background(), ethereum.CallMsg{
		To:   &t.contractAddr,
		Data: data,
	}, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to call allowance: %v", err)
	}

	// Unpack result
	allowance := new(big.Int)
	err = t.contractABI.UnpackIntoInterface(&[]interface{}{allowance}, "allowance", result)
	if err != nil {
		return nil, fmt.Errorf("failed to unpack allowance: %v", err)
	}

	// Convert to decimal
	decimals := big.NewInt(int64(TokenDecimals))
	divisor := new(big.Int).Exp(big.NewInt(10), decimals, nil)
	formatted := new(big.Float).Quo(new(big.Float).SetInt(allowance), new(big.Float).SetInt(divisor))

	return &Balance{
		Raw:       allowance,
		Formatted: formatted,
	}, nil
}

// Transfer sends tokens to another address
func (t *TragToken) Transfer(to string, amount *big.Float) (*types.Transaction, error) {
	if t.privateKey == nil {
		return nil, fmt.Errorf("private key required for transfers")
	}

	toAddr := common.HexToAddress(to)

	// Convert amount to wei
	decimals := big.NewInt(int64(TokenDecimals))
	multiplier := new(big.Int).Exp(big.NewInt(10), decimals, nil)
	amountWei := new(big.Int)
	amount.Mul(amount, new(big.Float).SetInt(multiplier)).Int(amountWei)

	// Get nonce
	nonce, err := t.client.PendingNonceAt(context.Background(), t.publicAddress)
	if err != nil {
		return nil, fmt.Errorf("failed to get nonce: %v", err)
	}

	// Get gas price
	gasPrice, err := t.client.SuggestGasPrice(context.Background())
	if err != nil {
		return nil, fmt.Errorf("failed to get gas price: %v", err)
	}

	// Create transaction
	auth, err := bind.NewKeyedTransactorWithChainID(t.privateKey, big.NewInt(56)) // BSC Chain ID
	if err != nil {
		return nil, fmt.Errorf("failed to create transactor: %v", err)
	}

	auth.Nonce = big.NewInt(int64(nonce))
	auth.Value = big.NewInt(0)
	auth.GasLimit = uint64(100000)
	auth.GasPrice = gasPrice

	// Pack transfer data
	data, err := t.contractABI.Pack("transfer", toAddr, amountWei)
	if err != nil {
		return nil, fmt.Errorf("failed to pack transfer data: %v", err)
	}

	// Create and send transaction
	tx := types.NewTransaction(nonce, t.contractAddr, big.NewInt(0), uint64(100000), gasPrice, data)
	signedTx, err := types.SignTx(tx, types.NewEIP155Signer(big.NewInt(56)), t.privateKey)
	if err != nil {
		return nil, fmt.Errorf("failed to sign transaction: %v", err)
	}

	err = t.client.SendTransaction(context.Background(), signedTx)
	if err != nil {
		return nil, fmt.Errorf("failed to send transaction: %v", err)
	}

	return signedTx, nil
}

// Helper method to build call message
func (t *TragToken) buildCallMsg(method string, args ...interface{}) ethereum.CallMsg {
	data, _ := t.contractABI.Pack(method, args...)
	return ethereum.CallMsg{
		To:   &t.contractAddr,
		Data: data,
	}
}

// Example usage
func main() {
	// Initialize without private key for read-only operations
	trag, err := NewTragToken("")
	if err != nil {
		log.Fatalf("Failed to initialize TRAG client: %v", err)
	}

	// Get token information
	fmt.Println("Token Info:")
	tokenInfo, err := trag.GetTokenInfo()
	if err != nil {
		log.Fatalf("Failed to get token info: %v", err)
	}

	fmt.Printf("  Name: %s\n", tokenInfo.Name)
	fmt.Printf("  Symbol: %s\n", tokenInfo.Symbol)
	fmt.Printf("  Decimals: %d\n", tokenInfo.Decimals)
	fmt.Printf("  Total Supply: %s\n", tokenInfo.TotalSupply.String())
	fmt.Printf("  Contract: %s\n", tokenInfo.ContractAddress)

	// Check balance
	fmt.Println("\nBalance Check:")
	balance, err := trag.GetBalance("0x742d35cc6564c0532d7e2d9db08d8d37b4b4e6c8")
	if err != nil {
		log.Fatalf("Failed to get balance: %v", err)
	}
	fmt.Printf("  Balance: %s TRAG\n", balance.Formatted.String())

	// For transactions, initialize with private key:
	// tragWithKey, err := NewTragToken("your_private_key_here")
	// tx, err := tragWithKey.Transfer("recipient_address", big.NewFloat(100))
}